src/grpc/zero_copy_stream_test.cc

```diff
@@ -166,68 +176,206 @@ TEST_F(GrpcZeroCopyInputStreamTest, Backups) {
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(5, size);
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(5, size);
   EXPECT_EQ(slice1.size() + slice2.size(),
             DelimiterToSize(reinterpret_cast<const unsigned char *>(data)));
 
   // Test the first slice
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(slice1.size(), size);
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(slice1.size(), size);
   EXPECT_EQ(slice1, std::string(reinterpret_cast<const char *>(data), size));
 
   // Back up & test again
   stream.BackUp(size);
   EXPECT_EQ(slice1.size() + slice2.size(), stream.BytesAvailable());
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(slice1.size(), size);
+  EXPECT_EQ(5, stream.ByteCount());
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(slice1.size(), size);
   EXPECT_EQ(slice1, std::string(reinterpret_cast<const char *>(data), size));
 
   // Now Back up 10 bytes & test again
   stream.BackUp(10);
   EXPECT_EQ(10 + slice2.size(), stream.BytesAvailable());
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(10, size);
+  EXPECT_EQ(5 + slice1.size() - 10, stream.ByteCount());
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(10, size);
   EXPECT_EQ(slice1.substr(slice1.size() - 10),
             std::string(reinterpret_cast<const char *>(data), size));
 
   // Test the second slice
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(slice2.size(), size);
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(slice2.size(), size);
   EXPECT_EQ(slice2, std::string(reinterpret_cast<const char *>(data), size));
 
   // Back up and test again
   stream.BackUp(size);
   EXPECT_EQ(slice2.size(), stream.BytesAvailable());
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(slice2.size(), size);
+  EXPECT_EQ(5 + slice1.size(), stream.ByteCount());
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(slice2.size(), size);
   EXPECT_EQ(slice2, std::string(reinterpret_cast<const char *>(data), size));
 
   // Now Back up size - 1 bytes (all but 1) and check again
   stream.BackUp(size - 1);
   EXPECT_EQ(slice2.size() - 1, stream.BytesAvailable());
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(slice2.size() - 1, size);
+  EXPECT_EQ(5 + slice1.size() + 1, stream.ByteCount());
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(slice2.size() - 1, size);
   EXPECT_EQ(slice2.substr(1),
             std::string(reinterpret_cast<const char *>(data), size));
 
   // Check the end of the stream
   EXPECT_FALSE(stream.Next(&data, &size));
 }
 
+TEST_F(GrpcZeroCopyInputStreamTest, Skips) {
+  GrpcZeroCopyInputStream stream;
+
+  // Create and add a message
+  std::string slice1 = "This is the first slice of the message.";
+  std::string slice2 = "This is the second slice of the message.";
+  std::string slice3 = "This is the third slice of the message.";
+  std::string slice4 = "This is the fourth slice of the message.";
+  const int kFrameSize = 5;
```

Comment:
Overall LGTM
Is it possible to put them into a test fixture and make the rest tests into several tests? It is hard to read while you need to keep a large context.

Reply:
This test is very stateful test; each test step depends on its position.  Even splitting into multiple test cases, each case still needs to move into that position in order to test that specific case, it will not help much for readability.

Reply:
Step sequencing is important in such tests, Backup, and Skip.  By splitting the test, the sequencing will be changed, will be simpler,  it will be different tests.  For now, I decided not to change it.